doc/design/scalar_switch_case_op_design.md

```diff
@@ -0,0 +1,75 @@
+### `scalar_switch_case_op` Design
+
+### Background
+In a general programing language, there are many scalar control flow operators such as `if` `else` `switch-case`, these operators take scalar bool values as condition input, then decide which block of code the program should run. PaddlePaddle fluid also needs such kind of operators.
+
+### For example:
+In learning rate decay, one strategy is called `piecewise_decay`, it has two inputs: `boundaries` and `values`.
+
+If the input parameters are:
+
+```
+boundaries = [100000, 110000]
+values = [1.0, 0.5, 0.1]
+```
+
+then the decay logic should be:
+
+```python
+def piecewise_decay()
+  learning_rate = init_lr
+
+  if step <= 100000:
+    learning_rate = 1.0
+  elif step > 100000 and step <= 110000:
+    learning_rate = 0.5
+  else:
+    learning_rate = 0.1
+
+  return learning_rate
+```
+
+### Solution
+To meet this kind of demand, we can have an `scalar_switch_case_op`:
+
+It's interface is:
+
+```python
+def piecewise_decay():
+  global_step = pd.Var(shape=[1], value=...)
+  learning_rate = pd.Var(shape=[1], value=...)
+
+  candition1 = pd.less_equal(global_step, 100000)
+  candition2 = pd.and(pd.less_then(100000, global_step), pd.less)
+
+  switch = ScalarSwitchOp()
+  with switch.case(candition1):
```

Comment:
do these three `case` similar to `if-else` or just `if`?  a little vague.

if it is like

```python
if cond1: pass
elif cond2: pass
else: pass #cond3
```
the three `switch.case` seems no logic relation with each other.

if it is just three `if` like

```python
if cond1: pass
if cond2: pass
if cond3: pass
```
then they shouldn't share the same `Switch` instance.

Need some improvement to specify what logic it is in first glimpse.

Reply:
yes, this is a problem, we have several condition and the relation is not clear, I will think about this.

Reply:
after some search, I think this situation can be more accurately described by `cond` op in Lisp <https://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node84.html>

![image](https://user-images.githubusercontent.com/3048612/35688763-dfc67690-07ac-11e8-95c9-f308edb7d180.png)

and the interface can be:
```python
def piecewise_decay():
  global_step = pd.Var(shape=[1], value=...)
  learning_rate = pd.Var(shape=[1], value=...)

  candition1 = pd.less_equal(global_step, 100000)
  candition2 = pd.and(pd.less_then(100000, global_step), pd.less)

  cand_op = ScalarCandOp()
  with cand_op.case(candition1):
    pd.assign(learning_rate, 1.0)
  with cand_op.case(candition2):
    pd.assign(learning_rate, 0.5)
  with cand_op.default_case():
    pd.assign(learning_rate, 0.1)

  return learning_rate
```

there are some features for `cond` op:
- 1. the order of condition is useful, the op will find a suitable branch to run from up to down in order.
- 2. only one branch will run.